[IMP] website, web_editor: redesign the grid layout and handle elements.

Brieuc-brd opened the pull request:
This PR improves the design of the grid layout, handle elements, overlay buttons, and drop zones to make them easier to
use.

task-3368793


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/132653).

Brieuc-brd commented:
Hey @stefanorigano 👋 When you have time, could you take a look at this PR ? Thanks !

stefanorigano reviewed (changes requested):
Thanks @Brieuc-brd , as said the visual result is really refreshing 👍 
Some comments for you in the code.

---

General note: there are a lot of negation selectors (`not()`) and it's usually an hint that the code is not well structured. 

Some suggestions:
In one circumstance you have set `&.readonly { display: none}`: to be tested but this rule may automatically invalidate others `&:not(.readonly)` rules because, since the element will not be visible anyway, it's not a big deal if we change it's design.

In other cases the negation could be avoided by creating a class.
```scss

.foobar {
  &.foo {
     [...]
  }

  &:not(.foo) {
     [...]
  }
}
```

```scss
.foobar {
   [common rules]
   
  &.foo {
     [...]
  }

  &.bar-not-foo { // new class that's never applied on foo
     [...]
  }
}
```

Since this PR aims to refactor handlers entirely, It's ok if we end-up with a bigger diff as long as we get cleaner and more maintainable code 👍 

---

Visual
---

- I find the "blue shape" that appears hovering "resizing" handlers unnecessary, the overall design would be much cleaner without it. 
This "shape" doesn't really add any value and doesn't specifically communicate that the behaviour is "resize' (in opposition to "padding", for example). Also, "padding" handlers are already different enough thanks to the stripes bg and the handler's button position 👍 

https://github.com/odoo/odoo/assets/8558351/227b36de-1612-40fc-9ab8-76244c2ea9ac

- When resizing an element the cursor temporarily switch to "default" until the snapping point is reached and/or the handler is hovered again. Nothing tells me that I'm currently dragging,  the behavior is confusing especially on "col mode" because there is not grid underneath showing the snapping points.
hint: when dragging there are classes applied to the `body` , you may find them useful ;)

https://github.com/odoo/odoo/assets/8558351/68999499-2487-4e72-b074-2a56666f562d

Brieuc-brd commented:
> I find the "blue shape" that appears hovering "resizing" handlers unnecessary, the overall design would be much cleaner without it.

I think it's necessary to add a visual element to indicate which edge we're handling. I've reworked the effect to make it more subtle.

> When resizing an element the cursor temporarily switch to "default" until the snapping point is reached and/or the handler is hovered again.

Fixed ✅ 

> General note: there are a lot of negation selectors (`not()`) and it's usually an hint that the code is not well structured.

I've reworked the code related to handles using more "grouping" classes to make the code easier to read.
The only tricky part using `:not()` is `body:not([class$="-important"])`: this prevents other edges from being hovered when you're currently triggering one.

Before :
https://github.com/odoo/odoo/assets/80679690/65dba1fe-7324-4ff0-9d5d-c23df3ee611b

After :
https://github.com/odoo/odoo/assets/80679690/6acadbe4-e601-4519-bf7b-8a6c429b2e6c

stefanorigano reviewed (changes requested):
Hello @Brieuc-brd , we're close, courage 💪

> I've reworked the effect to make it more subtle.

Accattable

> I've reworked the code related to handles using more "grouping" classes to make the code easier to read. 

Much better, thanks 👍 

> The only tricky part using :not() is body:not([class$="-important"])

Added a note in the code.


Don't forget to:
- update the PR's description
- check for trailing spaces: I still find a lot of them in your PRs, there are xcode's extensions to highlight these little *****  ;)

Brieuc-brd commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -2123,218 +2124,214 @@ we-select.o_we_border_preview_aligned_select {
-                &.e:after {
-                    content: "\f07e";
+                
+                body:not([class$="-important"]) &:hover, &:active {
```
I've updated the javascript part ✅ .

>About this specific SCSS, is there something missing in here? Not tested

This was intentional, if you already click on an element and then move the mouse, it will only trigger the hover state of other elements but not the active state.

Brieuc-brd commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -2055,19 +2055,20 @@ we-select.o_we_border_preview_aligned_select {
 @keyframes dropZoneInsert {
     to {
-        background-color: rgba($o-brand-odoo, 0.3);
+        background-color: rgba(mix($body-color, $o-we-handles-accent-color, 25%), 0.3);
```
I've reworked this part:
To get the correct value for `body-color`, I need to use `currentColor`. And so I had to use `color-mix()` function to play with it.
I also created a fallback variable for browsers that don't support `color-mix`.
What do you think ?

Brieuc-brd commented:
Hey @stefanorigano 👋 I updated the branch, could you take a look at this ?

>Don't forget to:
>
>update the PR's description

Done ✅ 

>check for trailing spaces: I still find a lot of them in your PRs, there are xcode's extensions to highlight these little ***** ;)

Done ✅

stefanorigano commented on addons/web_editor/static/src/scss/web_editor.variables.scss:
```diff
@@ -36,15 +36,18 @@ $o-we-ace-color: #2F3129 !default;
+$o-we-handle-inside-line-width: 3px !default;
 
-$o-we-dropzone-size: 30px !default; // $grid-gutter-width (todo: allow to use the variable)
+$o-we-dropzone-size: 60px !default; // $grid-gutter-width * 2 (todo: allow to use the variable)
```
`$grid-gutter-width` is actually `1.5rem (= 24px)` now.
Should this value be `48px` ?

Brieuc-brd replied:
In `website` this value is still 30px, I guess we should keep 60px.
https://github.com/odoo/odoo/blob/04eed39aa849aa25248906bd239fa396e90c72c0/addons/website/static/src/scss/bootstrap_overridden.scss#L77C23-L77C23

stefanorigano reviewed (changes requested):
Thanks @Brieuc-brd 🙌
Some notes for you

Brieuc-brd commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -2055,19 +2055,21 @@ we-select.o_we_border_preview_aligned_select {
+        width: auto;
+        border-radius: $border-radius-lg;
+        border: $o-we-dropzone-border-width dashed $o-we-dropzone-color-fallback;
+        border: $o-we-dropzone-border-width dashed $o-we-dropzone-color;
```
I've changed the approach by defining the colors in `website.scss` to be able to use `$body-bg` correctly. 
What do you think ?

Brieuc-brd commented on addons/web_editor/static/src/scss/web_editor.variables.scss:
```diff
@@ -36,15 +36,18 @@ $o-we-ace-color: #2F3129 !default;
 // has a page with a lot of content (10000px was proven to be too small)
 $o-we-handles-offset-to-hide: 100000px !default;
-$o-we-handles-btn-size: 20px !default;
+$o-we-handles-btn-size: 14px !default;
```
I'm not sure these elements should be the same size. The `oe_overlay` elements and the sidebar are two separate parts of the UI.
![Capture d’écran 2023-09-07 à 13 46 46](https://github.com/odoo/odoo/assets/80679690/4b0f2781-6918-4565-9db3-d65c875abc82)

In addition, the calculation to create the inner line must have an even value, otherwise rendering is not good.
![Capture d’écran 2023-09-07 à 13 53 37](https://github.com/odoo/odoo/assets/80679690/b59b6085-c11d-4f20-bb52-ebbcb5a037b9)

stefanorigano reviewed (approved):
Thanks @Brieuc-brd  :)

Brieuc-brd commented:
Hello @odoo/rd-website 👋 
This PR is ready for tech. review, could you take a look at it ? Thanks !

qsm-odoo commented:
Hello @Brieuc-brd 

Thanks for the work :muscle: We will have a look asap but unfortunately I will ask you to wait for our merge of the new OWL UI (max next week) as we are blocking all master task merge regarding the UI on our side too :grimacing: This may need some rework after rebase then, not sure.

I'll keep you posted :+1:

sobo-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5186,9 +5173,6 @@ registry.layout_column = SnippetOptionWidget.extend({
-        // Setting the z-index to the maximum of the grid.
-        gridUtils._setElementToMaxZindex(newColumnEl, rowEl);
-
         // Add the new column and update the grid height.
```
You should not remove these lines because without this, when we add a new element with the "Add element" option, it now goes behind the elements already in the grid instead of in front of them and so we don't see it directly anymore (only its overlay) :wink: 
This is also the part that adds the z-index property to these elements and so without that, it is not consistent

Brieuc-brd replied:
Done :)

sobo-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -1188,7 +1188,6 @@ var SnippetEditor = Widget.extend({
                     } else {
                         gridUtils._setElementToMaxZindex(self.$target[0], rowEl);
                     }
-                    gridUtils._setElementToMaxZindex(backgroundGridEl, rowEl);
```
The fact that the grid was placed in front of the elements was useful for cases where the grid would be so full we would not see it anymore (e.g. the Masonry and Big Boxes snippets, or any grid where the grid items would have a background) so this behavior should be restored :slightly_smiling_face: Even if there is the blue drag helper to see where to drop, the grid should still be visible and since it is now fully transparent, it should not "obstruct the view" anymore :slightly_smiling_face: 

Note that it should also be restored in the `snippets.options.js` file :)

Brieuc-brd replied:
I understand your point of view, but I'm not sure that the grid should always be in front of the elements 🤔 .

The grid is mainly an indicator to help you understand which mode you're in.
You don't need the grid to move your elements precisely, because your movements are already locked to the grid.
Also, if I'm not mistaken, "Big Boxes" is the only snippet where elements cover the entire snippet by default. You don't need the grid to understand that you can drag handles and edges to move and modify your elements.

> the grid should still be visible and since it is now fully transparent, it should not "obstruct the view" anymore

The grid is not fully transparent, it uses opacity but remains too visible when placed in front of elements. This was made on purpose to be sure that the grid remains visible in all cases, on the background image of a Banner for example 🙂 .

| Readable  | Less readable |
| ------------- | ------------- |
| ![Capture d’écran 2023-09-18 à 15 47 37 (2)](https://github.com/odoo/odoo/assets/80679690/d4ee4fe1-a267-4ea9-a6c6-489bd093eabe)  | ![Capture d’écran 2023-09-18 à 15 49 59 (2)](https://github.com/odoo/odoo/assets/80679690/b460a88c-2979-4484-b861-b737070e9f07)

By the way, Squarespace does exactly the same thing : the grid is behind the elements and is no longer visible when you cover it. This is probably done on purpose 🙂 

|  |  |
| ------------- | ------------- |
| ![Capture d’écran 2023-09-18 à 16 30 54](https://github.com/odoo/odoo/assets/80679690/2b632830-015f-45e4-ae27-269e7eef5c57)  | ![Capture d’écran 2023-09-18 à 15 46 25](https://github.com/odoo/odoo/assets/80679690/370e61b5-93ef-4ee8-8e17-d27db472a3a4)

But maybe I'm wrong 😅, what do you think @pierre-pde ? Do you prefer the grid in front of or behind elements ?

qsm-odoo replied:
Good point that it is the same behavior in squarespace but it still does not feel right to me either: if we don't need the grid (either functionally or aesthetically), why do we bother having it at all? In the Masonry snippet case, it feels not right to me either, there are blocks  covering the whole thing, you can move them and for some reason there is a piece of grid behind them... but same as in your squarespace screenshot I guess :grimacing: 

Decision for you @pierre-pde :)

qsm-odoo replied:
To be clear, I don't mind either way but: "show it everywhere" or "never show it" feels more consistent than "show it sometimes". But maybe this is not about consistency :smile:

pierre-pde replied:
@Brieuc-brd :
We had a quick chat with Soukeina:
1. I agree with Quentin & Soukeina that, because we have building blocks that cover entirely the grid (Masonry & Big boxes), it's probably **not** a good idea to do like Squarespace: the grid behind all elements.
2. We thought of the solution "put the grid in the second position, behind selected element". The problem is that the order is then not respected anymore. The selected element appears always on top, and might go back behind other element(s) once dropped.
3. Soukeina proposed the following comprise; keeping the grid above all elements, but putting the grid opacity from 0.25 to 0.1, it looks great IMO :) 
![image](https://github.com/odoo/odoo/assets/75257428/f7c12051-adc7-45f4-b58a-8d1dc555faf2)
What do you think?
cc @qsm-odoo

Brieuc-brd replied:
Indeed, it's easier to read than with an opacity of .25, and the grid still seems visible when there's a background image. Let's go with that, I'll update the branch 🙂

sobo-odoo reviewed (commented):
@Brieuc-brd Out of curiosity, I wanted to have a look and it looks really nice :slightly_smiling_face:  
But I noted some things that should be fixed (see review comments)
I asked @qsm-odoo before telling you :wink:

qsm-odoo commented:
@Brieuc-brd By the way: we decided to not merge the OWL conversion of the editor for 17.0. So once you take care of @sobo-odoo 's rightful comments, we'll deep test and merge this :+1:

Brieuc-brd commented:
Hey @odoo/rd-framework-editor 👋 
Since this PR is also related to the web editor, would you like to take a look at it ? 🙂

qsm-odoo commented:
@odoo/rd-framework-editor ( cc @Brieuc-brd ) I don't think this should interest you since it's not related to text edition at all :wink: I'll be looking at it now and merging it if it is fine. If you want to give opinions, you can still do it of course (even post-merge) :wink:

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -520,7 +520,7 @@ export class Wysiwyg extends Component {
 
         if (options.snippets) {
-            $(this.odooEditor.document.body).addClass('editor_enable');
+            $(this.odooEditor.document.body).addClass('editor_enable o_handlers_idle');
```
Instead of having to touch this line, could not you reverse the logic? (adding a class during the time it is not "idle"). Also could you not add it on the overlay instead of the body?

Brieuc-brd replied:
This was a solution we found with @stefanorigano (https://github.com/odoo/odoo/pull/132653#discussion_r1312895822) to avoid "ugly" css selector like this one : 
```css
 body:not([class$="-important"]) &:hover, &:active {
 ```
 What do you think ?

qsm-odoo replied:
This was indeed ugly but I think my two points remain the same:
- Right now you have: "there are styles for when o_handlers_idle is there and there are styles for o_handlers_idle is not there", you could just invert that logic: so in your example, you would not have `body:not([class$="-important"])` but `body:not(.o_handlers_enabled)` (something like that)
- If I am not mistaken, your style which depends on this `o_handlers_idle` class only affects the overlay: so could not the class be toggled on the overlay itself rather than on the body?

Brieuc-brd replied:
> "there are styles for when o_handlers_idle is there and there are styles for o_handlers_idle is not there"

There are styles on hover **only** when `o_handlers_idle` is there. The goal is to prevent `:hover` styles from other handles when the user is already dragging a handle. So I guess this logic makes sens, or did I missed something ? 🤔 

Done, I moved the class into the overlay 👍

qsm-odoo commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -2125,218 +2125,214 @@ we-select.o_we_border_preview_aligned_select {
-                    background-color: darken($o-we-handles-accent-color, 10%);
-                    color: white;
-                }
+                    width: $o-we-handles-btn-size;
```
Nitpicking: I prefer `width` before `height` in general (consistency with the rest, common practice) but in this case, it would also allow to not break the history of that line for no reason :wink:

Brieuc-brd replied:
Done 👍

qsm-odoo commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -2125,218 +2125,214 @@ we-select.o_we_border_preview_aligned_select {
-                &.readonly {
-                    cursor: auto !important;
+                    &.o_side_y {
+                        &:before {
```
To fix everywhere: `::before` and `::after` (`::`, not `:`). Even though `:` should work on all browsers, this is not technically correct. I know there were mistakes before but time to make cleaner code :muscle:

qsm-odoo commented on addons/website/static/src/scss/website.scss:
```diff
@@ -151,6 +151,10 @@ $-seen-urls: ();
     }
     @include print-variable('custom-colors', $custom-colors);
+
+    //7) Set colors for the drop zone
+    @include print-variable('dropzone-contrast-rgb', to-rgb(color-contrast($body-bg, $color-contrast-dark: darken($o-we-handles-accent-color, 25%))));
+    @include print-variable('dropzone-bg-color-rgb', to-rgb(mix($o-we-handles-accent-color, $body-bg, 50%)));
```
- `website.scss` is a visitor SCSS file, not an edit mode one. Even though most of the variable above are only used for edit mode, they technically make sense to be available for visitor pages. Those 2 new ones do not: they should only be available in edit mode SCSS files, in fact probably in the same place where you use them? `website.edit_mode.scss`?

- Although, why is it a website code? Should not it be in web_editor so that mass_mailing have the same design? I tested and there are indeed differences: already when you just drag and drop a snippet from the right panel.

- Speaking of dropdown, this is the only thing that I don't really like in the new design (I know @sobo-odoo had the same opinion but did not dare saying it out loud here :smile: Yes I changed my mind when seeing it in action @sobo-odoo). I don't mind removing the border and having a blurry effect, but they are multiple problems with those I think:
    - We don't see them as well, especially since they entirely disappear from the screen with their animation
    - When they go across each other, it's messy (always the case with inner snippets)
    - There is a strange (IMO) margin on both sides
    - Test drag and dropping fields in a "form" snippet -> this is broken :wink: 
    - Is there really a reason to have 2 kind of drop-zones in the end? :)

pierre-pde replied:
I agree that the new inner dropzone are not visible enough + tends to overlap too much.
Regarding the "section" dropzone, I think I prefer when they were more full width.
cc @qsm-odoo

qsm-odoo reviewed (changes requested):
Not tested deeply yet, but looks cool as already said :partying_face: I have some comments though, mostly about some dropzones :) Could use @pierre-pde advice too I guess :)

Brieuc-brd commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4682,19 +4678,11 @@ registry.sizing = SnippetOptionWidget.extend({
             var $handle = $(handle);
             var direction = $handle.hasClass('n') ? 'top' : 'bottom';
-            $handle.height(self.$target.css('padding-' + direction));
+            $handle.outerHeight(self.$target.css('padding-' + direction));
```
When `.height` is used, the handle area overlaps the content according to the size of its border (here, 2px).
Using `.outerHeight` solves this issue.
| `.height`  | `.outerHeight` |
| ------------- | ------------- |
| ![Capture d’écran 2023-09-21 à 15 01 39](https://github.com/odoo/odoo/assets/80679690/20fa1759-ce6a-4f01-8ab9-d0d20df1dd2c)  | ![Capture d’écran 2023-09-21 à 15 07 06](https://github.com/odoo/odoo/assets/80679690/4c3e69e2-0a32-4595-99b1-8b8034234822)  |

qsm-odoo replied:
Well spotted :+1: But then I think I am seeing the same thing in stable version :fearful: For a one line change, I would have considered merging that in stable.

But let's say we should not because it is not a critical problem. However, what you should always try to do is making that an independent commit. So, here I'd say: make this a dedicated commit in this same PR (that you can label with [FIX]). This is important because I am asking this question now but if I had not seen that you made this change and were to wonder "why `outerHeight`?" in two years, I would blame that line and found out your [IMP] commit which does not explain why we had to use `outerHeight` instead of height :wink:

Brieuc-brd replied:
Done 👍 f9a603399349a36f7ab701ced19a341eada0850e

pierre-pde commented:
@Brieuc-brd , I found one potential imp & one bug :) 
![image](https://github.com/odoo/odoo/assets/75257428/535a18cd-0e25-420b-be69-57bb307d4f34)

Brieuc-brd commented:
Hey @pierre-pde 👋 
Point 1: I'm not sure it's related to this PR but I can take a look 👍 
Point 2: this issue is already in saas-16.2, I think it should be fixed in another PR.

Brieuc-brd commented:
@pierre-pde 
I added a commit to change the style of `o_editable` elements on hover, I used solid borders because dashed borders are more related to drop zones 🙂 .

![Capture d’écran 2023-09-25 à 16 22 11](https://github.com/odoo/odoo/assets/80679690/3c7b57c0-e403-4d64-b642-3009bc3ed39f)

pierre-pde commented:
Thanks @Brieuc-brd , I'll create a task for point 2

qsm-odoo reviewed (approved):
Great job as always :+1: I like merging negative diff for prettier results :smile: 
I made [this small adaptation](https://github.com/odoo/odoo/compare/93a24b0cd9e04c0d84a523cdc95a696d1e127dfb..7c5b6ff92207f8e67e36ea83f71d193efd47bdbb) (really, details + nitpicking to force my own coding style).

@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.
